File: hash.cpp
#include "hash.h"
#include "serialize.h"
#include <iostream>
#include <cmath>

using namespace std;

auto hash_str(const string& key, int mod) -> int
{
    long long hashVal = 0;
    for(char character: key)
    {
        hashVal = (hashVal * 31 + std::abs((int)character)) % mod;
    }
    return (int)hashVal;
}

ChainHash::ChainHash(int buckCount) : table(nullptr), cap(buckCount)
{
    if(cap < 1) cap=1;
    table = new Node*[cap];
    for(int i = 0; i< cap; i++) table[i] = nullptr;
}

ChainHash::~ChainHash()
{
    for(int i=0; i < cap; i++)
    {
        Node* cur = table[i];
        while(cur != nullptr)
        {
            Node* nxt = cur->next;
            delete cur;
            cur = nxt;
        }
    }
    delete[] table;
    table = nullptr;
    cap = 0;
}

void ChainHash::insert(const string& key, const string& value) {
    int idx = hash_str(key, cap);

    Node* head = table[idx];
    Node* cur = head;
    while (cur != nullptr) {
        if (cur->key == key) {
            cur->val = value;
            return;
        }
        cur = cur->next;
    }

    Node* newNode = new Node(key, value);
    newNode->next = head;
    table[idx] = newNode;
}

auto ChainHash::erase(const string& key) -> bool
{
    int idx = hash_str(key, cap);

    Node* cur = table[idx];
    Node* prv = nullptr;

    while(cur != nullptr)
    {
        if(cur->key == key)
        {
            if(prv != nullptr) {
                prv->next = cur->next;
            } else{
                table[idx] = cur->next;
            }
            delete cur;
            return true;
        }
        prv = cur;
        cur = cur->next;
    }
    return false;
}

auto ChainHash::find(const string& key) -> string
{
    int idx = hash_str(key, cap);
    Node* cur = table[idx];

    while (cur != nullptr) {
        if (cur->key == key) {
            return cur->val;
        }
        cur = cur->next;
    }
    return "";
}

void ChainHash::show() {
    cout << "цепочки: ";
    bool any = false;
    for (int i = 0; i < cap; i++) {
        if (table[i] != nullptr) {
            any = true;
            cout << "[" << i << "]" << ": ";
            Node* cur = table[i];
            while (cur != nullptr) {
                cout << cur->key << "->" << cur->val;
                if (cur->next != nullptr) cout << " -> ";
                cur = cur->next;
            }
            cout << "  ";
        }
    }
    if (!any) cout << "пусто";
    cout << "\n";
}

void ChainHash::clearAndResize(int newCap) {
    for(int i = 0; i < cap; i++) {
        Node* cur = table[i];
        while(cur != nullptr) {
            Node* nxt = cur->next;
            delete cur;
            cur = nxt;
        }
    }
    delete[] table;
    cap = newCap;
    table = new Node*[cap];
    for(int i = 0; i < cap; i++) table[i] = nullptr;
}

OpenHash::OpenHash(int size) : table(nullptr), cap(size)
{
    if(cap < 1) cap=1;
    table = new Entry[cap];
    for (int i = 0; i < cap; ++i) {
        table[i].used    = false;
        table[i].deleted = false;
    }
}

OpenHash::~OpenHash()
{
    delete[] table;
    table = nullptr;
    cap = 0;
}

void OpenHash::insert(const string& key, const string& value)
{
    int idx = hash_str(key, cap);
    int start = idx;

    while(true)
    {
        if(!table[idx].used && !table[idx].deleted)
        {
            table[idx].key = key;
            table[idx].val = value;
            table[idx].used = true;
            table[idx].deleted = false;
            return;
        }
        if (table[idx].used && table[idx].key == key) {
            table[idx].val = value;
            return;
        }
        if (table[idx].deleted) {
            table[idx].key     = key;
            table[idx].val     = value;
            table[idx].used    = true;
            table[idx].deleted = false;
            return;
        }

        idx = (idx+1) % cap;

        if(idx == start)
        {
            cout << "Таблица переполнена" << "\n";
            return;
        }
    }
}

auto OpenHash::erase(const string& key) -> bool
{
    int idx = hash_str(key, cap);
    int start = idx;

    while(true)
    {
        if (table[idx].used && table[idx].key == key) {
            table[idx].used    = false;
            table[idx].deleted = true;
            return true;
        }
        if (!table[idx].used && !table[idx].deleted) {
            return false;
        }
        idx = (idx + 1) % cap;

        if (idx == start) {
            return false;
        }
    }
}

auto OpenHash::find(const string& key) const -> string {
    int idx   = hash_str(key, cap);
    int start = idx;

    while (true) {
        if (table[idx].used && table[idx].key == key) {
            return table[idx].val;
        }
        if (!table[idx].used && !table[idx].deleted) {
            return "";
        }
        idx = (idx + 1) % cap;
        if (idx == start) {
            return "";
        }
    }
}

void OpenHash::show() {
    cout << "ОТКР.АДР: ";
    bool any = false;
    for (int i = 0; i < cap; i++) {
        if (table[i].used) {
            any = true;
            cout << "[" << i << "]" << table[i].key << "->" << table[i].val << "  ";
        } else if (table[i].deleted) {
            any = true;
            cout << "[" << i << "]" << "(удалено)  ";
        }
    }
    if (!any) cout << "пусто";
    cout << "\n";
}

void OpenHash::clearAndResize(int newCap) {
    delete[] table;
    cap = newCap;
    table = new Entry[cap];
    for (int i = 0; i < cap; ++i) {
        table[i].used = false;
        table[i].deleted = false;
    }
}

auto is_space_char(char charToCheck) -> bool {
    return charToCheck == ' ' || charToCheck == '\t';
}

void parse_cmd(const string& line, string& cmd, string& arg1, string& rest) {
    cmd.clear(); arg1.clear(); rest.clear();
    int lineLength = (int)line.size();
    int idx = 0;
    while (idx < lineLength && is_space_char(line[idx])) { idx++; }

    while (idx < lineLength && !is_space_char(line[idx])) {
        cmd.push_back(line[idx]); idx++;
    }
    while (idx < lineLength && is_space_char(line[idx])) { idx++; }
    while (idx < lineLength && !is_space_char(line[idx])) {
        arg1.push_back(line[idx]); idx++;
    }
    while (idx < lineLength && is_space_char(line[idx])) { idx++; }
    if (idx < lineLength) rest = line.substr(idx);
}

void hash_man() {
    cout << "УПРАВЛЕНИЕ ХЕШ-ТАБЛИЦАМИ" << "\n";
    ChainHash htChain(10);
    OpenHash  htOpen(10);

    while (true) {
        cout << "\n";
        cout << "Команды: INSERT k v | DELETE k | SEARCH k | SHOW | BACK" << "\n";
        cout << "Введите команду: ";
        string line;
        if (!getline(cin, line)) { break; }
        string cmd, arg1, rest;
        parse_cmd(line, cmd, arg1, rest);

        if (cmd.empty()) continue;
        if (cmd == "BACK") {
            cout << "Выход" << endl;
            break;
        }
        if (cmd == "SHOW") {
            htChain.show();
            htOpen.show();
            continue;
        }
        if (cmd == "INSERT") {
            htChain.insert(arg1, rest);
            htOpen.insert(arg1, rest);
            continue;
        }
        if (cmd == "DELETE") {
            htChain.erase(arg1);
            htOpen.erase(arg1);
            continue;
        }
        if (cmd == "SEARCH") {
            string v1 = htChain.find(arg1);
            string v2 = htOpen.find(arg1);
            cout << "Chain: " << (v1.empty() ? "Not found" : v1) << "\n";
            cout << "Open:  " << (v2.empty() ? "Not found" : v2) << "\n";
            continue;
        }
    }
}

File: ds/hash.go
package ds

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func hashStr(key string, mod int) int {
	h := 0
	for _, ch := range key {
		v := int(ch)
		if v < 0 {
			v = -v
		}
		h = (h*31 + v) % mod
	}
	return h
}

type ChainNode struct {
	Key  string
	Val  string
	Next *ChainNode
}

type ChainHash struct {
	Table []*ChainNode
	Cap   int
}

func NewChainHash(buckets int) *ChainHash {
	if buckets < 1 {
		buckets = 1
	}
	return &ChainHash{
		Table: make([]*ChainNode, buckets),
		Cap:   buckets,
	}
}

func (c *ChainHash) Insert(k, v string) {
	idx := hashStr(k, c.Cap)
	cur := c.Table[idx]

	for cur != nil {
		if cur.Key == k {
			cur.Val = v
			return
		}
		cur = cur.Next
	}

	c.Table[idx] = &ChainNode{
		Key:  k,
		Val:  v,
		Next: c.Table[idx],
	}
}

func (c *ChainHash) Find(key string) string {
	idx := hashStr(key, c.Cap)
	cur := c.Table[idx]
	for cur != nil {
		if cur.Key == key {
			return cur.Val
		}
		cur = cur.Next
	}
	return ""
}

func (c *ChainHash) Erase(key string) bool {
	idx := hashStr(key, c.Cap)
	cur := c.Table[idx]
	var prev *ChainNode = nil

	for cur != nil {
		if cur.Key == key {
			if prev == nil {
				c.Table[idx] = cur.Next
			} else {
				prev.Next = cur.Next
			}
			return true
		}
		prev = cur
		cur = cur.Next
	}
	return false
}

func (c *ChainHash) Show() {
	fmt.Println("--- ChainHash Content ---")
	for i, node := range c.Table {
		if node != nil {
			fmt.Printf("[%d]: ", i)
			for cur := node; cur != nil; cur = cur.Next {
				fmt.Printf("(%s->%s) -> ", cur.Key, cur.Val)
			}
			fmt.Println("nil")
		}
	}
}

type OpenEntry struct {
	Key     string
	Val     string
	Used    bool
	Deleted bool
}

type OpenHash struct {
	Table []OpenEntry
	Cap   int
}

func NewOpenHash(size int) *OpenHash {
	if size < 1 {
		size = 1
	}
	return &OpenHash{
		Table: make([]OpenEntry, size),
		Cap:   size,
	}
}

func (o *OpenHash) Insert(k, v string) {
	idx := hashStr(k, o.Cap)
	start := idx

	for {
		e := &o.Table[idx]

		if !e.Used && !e.Deleted {
			e.Key = k
			e.Val = v
			e.Used = true
			return
		}
		if e.Used && e.Key == k {
			e.Val = v
			return
		}
		if e.Deleted {
			e.Key = k
			e.Val = v
			e.Used = true
			e.Deleted = false
			return
		}

		idx = (idx + 1) % o.Cap
		if idx == start {
			fmt.Println("OpenHash full!")
			return
		}
	}
}

func (o *OpenHash) Find(k string) string {
	idx := hashStr(k, o.Cap)
	start := idx

	for {
		e := &o.Table[idx]

		if e.Used && e.Key == k {
			return e.Val
		}
		if !e.Used && !e.Deleted {
			return ""
		}

		idx = (idx + 1) % o.Cap
		if idx == start {
			return ""
		}
	}
}

func (o *OpenHash) Erase(k string) bool {
	idx := hashStr(k, o.Cap)
	start := idx

	for {
		e := &o.Table[idx]

		if e.Used && e.Key == k {
			e.Used = false
			e.Deleted = true
			return true
		}

		if !e.Used && !e.Deleted {
			return false
		}

		idx = (idx + 1) % o.Cap
		if idx == start {
			return false
		}
	}
}

func (o *OpenHash) Show() {
	fmt.Println("--- OpenHash Content ---")
	for i, e := range o.Table {
		status := "EMPTY"
		if e.Used {
			status = fmt.Sprintf("%s->%s", e.Key, e.Val)
		} else if e.Deleted {
			status = "DELETED"
		}
		fmt.Printf("[%d]: %s\n", i, status)
	}
}

func HashMan() {
	scanner := bufio.NewScanner(os.Stdin)
	ch := NewChainHash(10)
	oh := NewOpenHash(10)

	currentMode := "chain"

	fmt.Println("Хеш-менеджер. Команды: CHAIN, OPEN, INSERT k v, SEARCH k, DELETE k, SHOW, BACK")

	for {
		fmt.Printf("[%s] >> ", currentMode)
		if !scanner.Scan() {
			break
		}
		line := scanner.Text()
		parts := strings.Fields(line)
		if len(parts) == 0 {
			continue
		}

		cmd := parts[0]

		switch cmd {
		case "BACK":
			return
		case "CHAIN":
			currentMode = "chain"
			fmt.Println("Переключено на ChainHash")
		case "OPEN":
			currentMode = "open"
			fmt.Println("Переключено на OpenHash")
		case "SHOW":
			if currentMode == "chain" {
				ch.Show()
			} else {
				oh.Show()
			}
		case "INSERT":
			if len(parts) < 3 {
				fmt.Println("Нужен ключ и значение")
				continue
			}
			if currentMode == "chain" {
				ch.Insert(parts[1], parts[2])
			} else {
				oh.Insert(parts[1], parts[2])
			}
		case "SEARCH":
			if len(parts) < 2 {
				fmt.Println("Нужен ключ")
				continue
			}
			res := ""
			if currentMode == "chain" {
				res = ch.Find(parts[1])
			} else {
				res = oh.Find(parts[1])
			}
			if res == "" {
				fmt.Println("Не найдено")
			} else {
				fmt.Println("Найдено:", res)
			}
		case "DELETE":
			if len(parts) < 2 {
				fmt.Println("Нужен ключ")
				continue
			}
			ok := false
			if currentMode == "chain" {
				ok = ch.Erase(parts[1])
			} else {
				ok = oh.Erase(parts[1])
			}
			if ok {
				fmt.Println("Удалено")
			} else {
				fmt.Println("Не найдено для удаления")
			}
		default:
			fmt.Println("Неизвестная команда")
		}
	}
}